File: backtracking/46. Permutations.js
`Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.
Example 1:

Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
Example 2:

Input: nums = [0,1]
Output: [[0,1],[1,0]]
Example 3:

Input: nums = [1]
Output: [[1]]


`;

const permute = (nums) => {
  const res = [];
  //! base base
  if (nums.length === 1) {
    const cpy = [...nums];
    console.log(cpy);
    return [cpy];
  }
  //! recursion
  for (let i = 0; i < nums.length; i++) {
    const topVal = nums.shift();
    const subRes = permute(nums); // subRes is an array of arrays which is the result of permute(nums)
    for (let num of subRes) {
      num.push(topVal);
    }
    console.log({ subRes });
    res.push(...subRes);
    nums.push(topVal);
  }
  return res;
};

console.log(permute([1, 2, 3]));

File: backtracking/46. Permutations.py
class Solution:
    def permute(self, nums):
        res = []
        self.backtrack(nums, [], res)
        print(res)
        return res

    def backtrack(self, nums, path, res):
        if not nums:
            res.append(path)
            return

        for i in range(len(nums)):
            self.backtrack(nums[:i] + nums[i+1:], path + [nums[i]], res)


# Path: backtracking/47. Permutations II.py

if __name__ == "__main__":
    nums = [1,  2, 3]
    Solution().permute(nums)
